lldb/tools/lldb-server/lldb-platform.cpp

```diff
@@ -421,97 +417,210 @@ int main_platform(int argc, char *argv[]) {
+                Status error = spawn_process(
+                    progname, sock_up.get(), gdbserver_port, port_offset,
+                    inferior_arguments, log_file, log_channels);
+                if (error.Fail()) {
+                  Log *log = GetLog(LLDBLog::Platform);
+                  LLDB_LOGF(log, "spawn_process failed: %s", error.AsCString());
+                  WithColor::error()
+                      << "spawn_process failed: " << error.AsCString() << "\n";
+                  if (!g_server) {
+                    g_terminate = true;
+                    g_main_loop->AddPendingCallback(
+                        [](MainLoopBase &loop) { loop.RequestTermination(); });
+                  }
+                }
+              }
+              sock_up.reset();
+            });
+    if (!platform_handles) {
+      printf("Failed to accept platform: %s\n",
+             llvm::toString(platform_handles.takeError()).c_str());
+      return socket_error;
     }
-    printf("Connection established.\n");
 
-    if (g_server) {
-      std::optional<uint16_t> available_port;
-      {
-        std::lock_guard<std::mutex> guard(gdbserver_portmap_mutex);
-        auto port = gdbserver_portmap.GetNextAvailablePort();
-        if (port)
-          available_port = *port;
-        else
-          llvm::consumeError(port.takeError());
-      }
-      if (!available_port) {
-        fprintf(stderr,
-                "no available gdbserver port for connection - dropping...\n");
-      } else {
-        const Socket *conn_socket =
-            static_cast<const Socket *>(conn->GetReadObject().get());
-        error =
-            spawn_process(progname, conn_socket, *available_port, port_offset,
-                          inferior_arguments, log_file, log_channels);
-        if (error.Fail()) {
-          {
-
-            std::lock_guard<std::mutex> guard(gdbserver_portmap_mutex);
-            gdbserver_portmap.FreePort(*available_port);
-          }
-          LLDB_LOGF(GetLog(LLDBLog::Platform), "spawn_process failed: %s",
-                    error.AsCString());
-          WithColor::error()
-              << "spawn_process failed: " << error.AsCString() << "\n";
-        }
-      }
-      // Parent doesn't need a connection to the lldb client
-      delete conn;
-
-      // Parent will continue to listen for new connections.
-      continue;
-    } else {
-      // If not running as a server, this process will not accept
-      // connections while a connection is active.
-      acceptor_up.reset();
-
-      // When not running in server mode, use all available ports
-      platform.SetPortMap(std::move(gdbserver_portmap));
+    llvm::Expected<std::vector<MainLoopBase::ReadHandleUP>> gdb_handles =
+        sock_gdb->Accept(*g_main_loop, [&platform, &gdb_inferior_arguments](
+                                           std::unique_ptr<TCPSocket> sock_up) {
+          printf("Gdb connection established.\n");
```

Comment:
Do we want to print this for every new connection?

Reply:
Removed.